### DISCUSSION TITLE
How to type the zod schema object based on a predefined type?

### DISCUSSION THREAD

**Discussion by samislam** [colinhacks/zod#3090] [Q&A]
I have the following code:
```ts
class MyModel {
  propA: string
  propB: number
  propC: boolean
  constructor() {
    this.propA = ''
    this.propB = 123
    this.propC = true
  }
}
```

I want to create a Zod schema object based on this class, so that the **keys** and the **values** will correspond to this type:
```ts
const mySchema = z.object<SOME_GENERIC_METHOD_HERE<MyModel>>({
propA: z.string(),
  propB: z.number(),
  propC: z
    .boolean()
    .optional()
    .transform((i) => i),
})
})
```

I have tried to create a simple generic function that worked perfectly when `strictNullChecks` option in the **tsconfig.json** is disabled:

```ts
type ZodSchemaKeysFrom1<E> = {
  [k in keyof E]: ZodType<E[k]>
}

const mySchema = z.object<ZodSchemaKeysFrom1<MyModel>>({
  propA: z.string(),
  propB: z.number(),
  propC: z
    .boolean()
    .optional()
    .transform((i) => i),
})

type MySchemaFields = z.infer<typeof mySchema>
```

In this case `MySchemaFields` type is being inferred correctly as:
```ts
type MySchemaFields = {
    propA?: string;
    propB?: number;
    propC?: boolean;
}
```
It's the same type that will be inferred if you remove the generic angle brackets from `z.object<xxx>()`

Now, the problem is, I don't want to disable this typescript flag, I want my code to be as  strict as possible to avoid future mistakes.

So when I remove this flag (`strictNullChecks`), I get the following error only on the propC in the zod schema:

[![enter image description here][1]][1]

The error messsage is:

```
Type 'ZodEffects<ZodOptional<ZodBoolean>, boolean | undefined, boolean | undefined>' is not assignable to type 'ZodType<boolean, ZodTypeDef, boolean>'.
  Types of property '_type' are incompatible.
    Type 'boolean | undefined' is not assignable to type 'boolean'.
      Type 'undefined' is not assignable to type 'boolean'.ts(2322)
index.ts(6, 3): The expected type comes from property 'propC' which is declared here on type 'ZodSchemaKeysFrom1<MyModel>'

(property) propC: ZodType<boolean, z.ZodTypeDef, boolean>

```

How can I avoid this error?

I have created a code sandbox for everyone to try it out.

[Click here to see and try it in my devbox codesandbox][2]

  [1]: https://i.stack.imgur.com/vXAqY.png
  [2]: https://codesandbox.io/p/devbox/zod-keys-schema-9krk5v?file=%2Fsrc%2Findex.ts

**Comment by JacobWeisenburger** [comment]
To the best of my knowledge there isn't a way to do this currently. However maybe someone else knows something that I don't.

**Comment by rydonahue1** [comment]
Came here looking for the same information. So far I am able to get what you're looking for by using

```
const MySchema = z.object({
   id: z.string()
}) as z.Schema<MyType>
```

However even though it throws an error if your zod schema doesnt match the type, it removes the ability to extend the schema which ruins the usefulness of this.

**Reply by JacobWeisenburger** [reply]
This is probably what you are looking for:

```ts
const MySchema = z.object( {
    id: z.string()
} ) satisfies z.Schema<MyType>
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by samislam** [reply]
> Came here looking for the same information. So far I am able to get what you're looking for by using
> 
> ```
> const MySchema = z.object({
>    id: z.string()
> }) as z.Schema<MyType>
> ```
> 
> However even though it throws an error if your zod schema doesnt match the type, it removes the ability to extend the schema which ruins the usefulness of this.


Instead of using "as", using "satisfies" will remove the drawback.

**Reply by rydonahue1** [reply]
Yes that works, thanks!

**Comment by samislam** [comment]
I found a small solution, but this will only typecheck the object property names, and will not care about the values.

I defined this custom generic type:
```ts
export type ZodKeysFrom<T> = {
  [K in keyof T]: unknown
}
```
And I am using it like this:
```ts
type MySchemaFields = {
    propA?: string;
    propB?: number;
    propC?: boolean;
}

const mySchema = z.object({
  propA: z.string(),
  propB: z.number(),
  propC: z
    .boolean()
    .optional()
    .transform((i) => i),
} satisfies ZodKeysFrom<MySchemaFields>)

type MySchemaFields = z.infer<typeof mySchema>
```
